iocage/lib/Jail.py

```diff
@@ -0,0 +1,500 @@
+            jail=self
+        )
+
+        command = [
+            "/sbin/route",
+            "add"
+        ] + (["-6"] if (ipv6 is True) else []) + [
+            "default",
+            gateway
+        ]
+
+        self.exec(command)
+
+    def require_jail_not_existing(self):
+        if self.exists:
+            msg = f"Jail {self.humanreadable_name} already exists"
+            self.logger.error(msg)
+            raise Exception(msg)
+
+    def require_jail_existing(self):
+        if not self.exists:
+            msg = f"Jail {self.humanreadable_name} does not exist"
+            self.logger.error(msg)
+            raise Exception(msg)
+
+    def require_jail_stopped(self):
+        if self.running:
+            msg = f"Jail {self.humanreadable_name} is already running"
+            self.logger.error(msg)
+            raise Exception(msg)
+
+    def require_jail_running(self):
+        if not self.running:
+            msg = f"Jail {self.humanreadable_name} is not running"
+            self.logger.error(msg)
+            raise Exception(msg)
+
+    def update_jail_state(self):
+        try:
+            stdout = subprocess.check_output([
+                "/usr/sbin/jls",
+                "-j",
+                self.identifier,
+                "-v",
+                "-h"
+            ], shell=False, stderr=subprocess.DEVNULL)
+            output = stdout.decode("utf-8").strip()
+
+            keys, values = [x.split(" ") for x in output.split("\n")]
+            self.jail_state = dict(zip(keys, values))
+
+        except:
+            self.jail_state = None
+
+    def _resolve_uuid(self, text):
+        jails_dataset = self.host.datasets.jails
+        for dataset in list(jails_dataset.children):
+            dataset_uuid = dataset.name[(len(jails_dataset.name) + 1):]
+            if text in dataset_uuid:
+                self.logger.debug(f"Resolved {text} to uuid {dataset_uuid}")
+                return dataset_uuid
+
+        raise Exception(f"No jail matching {text} was found")
+
+    def _get_name(self):
+        return self.humanreadable_name
+
+    def _get_humanreadable_name(self):
+
+        try:
+            return self.config.name
+        except:
+            pass
+
+        try:
+            return str(self.config.uuid)[:8]
+        except:
+            pass
+
+        raise Exception("This Jail does not have any identifier yet")
+
+    def _get_stopped(self):
+        return self.running is not True
+
+    def _get_running(self):
+        return self._get_jid() is not None
+
+    def _get_jid(self):
+        try:
+            return self.jail_state["jid"]
+        except:
+            pass
+
+        try:
+            self.update_jail_state()
+            return self.jail_state["jid"]
+        except:
+            return None
+
+    def _get_identifier(self):
+        return f"ioc-{self.uuid}"
+
+    def _get_exists(self):
+        try:
+            self.dataset
```

Comment:
In this function we want to know whether the dataset exists or not. When accessing the dataset attribute, the ZFS dataset is returned. If it does not exist an exception is raised by the property getter `dataset`. We want to return `False` if this was the case.